File: Zad2/Zadanie2.1.py
from typing import List
import random

class Klawiatura:

    shared = None

    class __Klawiatura:

         keyPressed = None 

         def __init__(self):
             pass

    def __init__(self):
        if not Klawiatura.shared:
            Klawiatura.shared = Klawiatura.__Klawiatura()

class Klawisz:
    value: int = 0
    __wasPressed: bool = False
    
    def __init__(self, keyValue):
        self.value = keyValue

    def checkIfPressed(self, klawiatura: Klawiatura):
        if not self.__wasPressed:
            print("Sprawdzenie czy klawisz {} zostal wcisniety\n".format(self.value))
            if self.value == klawiatura.keyPressed.value:
                print("Klawisz {} zostal wcisniety\n".format(self.value))
                self.__wasPressed = True
                return self.__wasPressed
            else:
                return self.__wasPressed
        else:
            return self.__wasPressed

def areAllKeysPressed(klawisze: List[Klawisz], klawiatura: Klawiatura):
    areAllKeysPressed: bool = True
    for klawisz in klawisze:
        if not klawisz.checkIfPressed(klawiatura):
            areAllKeysPressed = False
    return areAllKeysPressed

if __name__ == '__main__':
    klawiatura = Klawiatura().shared
    klawiatura.keyPressed = Klawisz(0)

    k1 = Klawisz(18)
    k2 = Klawisz(45)
    k3 = Klawisz(4)
    k4 = Klawisz(23)
    k5 = Klawisz(1)

    klawisze = [k1, k2, k3, k4, k5]
    counter = 0
    while(not areAllKeysPressed(klawisze, klawiatura)):
        klawiatura.keyPressed = Klawisz(random.randint(1, 50))
        print("\nIteracja {} \nWcisniety klawisz: {} \n".format(counter, klawiatura.keyPressed.value))
        counter += 1


    # print(klawiatura)
    # klawiatura.keyPressed = Klawisz(233223)
    # print(str(klawiatura.keyPressed.value))

    # klaw2 = Klawiatura().shared
    # print(klaw2)
    # print(Klawiatura().shared.keyPressed)

File: Zad2/Zadanie2_1.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

class Klawiatura
{
  private:
    static Klawiatura *shared;
    Klawiatura()
    {
        keyPressed = 0;
    }

  public:
    unsigned keyPressed;
    static Klawiatura *getInstance()
    {
        {
            if (shared == NULL)
            {
                shared = new Klawiatura();
                return shared;
            }
            else
            {
                return shared;
            }
        }
    }
};

class Klawisz
{
  private:
    bool wasPressed;

  public:
    unsigned key;

    Klawisz(unsigned key)
    {
        this->wasPressed = false;
        this->key = key;
    }

    bool checkIfPressed(Klawiatura keyboard)
    {
        if (!wasPressed)
        {
            cout << "Sprawdzenie czy klawisz " << key << " zostal wcisniety"
                 << "\n";
            if (key == keyboard.keyPressed)
            {
                cout << "Klawisz  " << key << " zostal wcisniety"
                     << "\n";
                wasPressed = true;
                return wasPressed;
            }
            else
            {
                return wasPressed;
            }
        }
        else
        {
            return wasPressed;
        }
    }
};

bool areAllKeysPressed(Klawisz klawisze[], int size, Klawiatura keyboard)
{
    bool allKeysPressed = true;
    for (int i = 0; i < size; i++)
    {
        if (!klawisze[i].checkIfPressed(keyboard))
        {
            allKeysPressed = false;
        }
    }
    return allKeysPressed;
}

Klawiatura *Klawiatura::shared(0);
int main()
{
    Klawiatura keyboard = *Klawiatura::getInstance();
    Klawisz k1 = Klawisz(18);
    Klawisz k2 = Klawisz(45);
    Klawisz k3 = Klawisz(4);
    Klawisz k4 = Klawisz(23);
    Klawisz k5 = Klawisz(1);

    const int arrayLength = 5;
    Klawisz klawisze[arrayLength] = {k1, k2, k3, k4, k5};

    int counter = 0;
    while (!areAllKeysPressed(klawisze, arrayLength, keyboard))
    {
        keyboard.keyPressed = rand() % 50 + 1;
        cout << "\n\nIteracja " << counter << "\n"
             << "Wcisniety klawisz: " << keyboard.keyPressed << "\n";
        counter++;
    }
    return 0;
}
